libclamav/clamav.h

```diff
@@ -601,6 +601,63 @@ enum cl_msg {
     CL_MSG_ERROR        = 128 /* LibClamAV ERROR: */
 };
 
+/**
+ * @brief Progress callback for sig-load, engine-compile, and engine-free.
+ *
+ * Disclaimer: total_items may increase with each call as new items to be
+ * processed are found. As such, total_items == now_completed does not mean it
+ * guaranteed to be complete.
+ *
+ * @param total_items   Total number of items
+ * @param now_completed Number of items completed
+ * @param context       Opaque application provided data
+ */
+typedef void (*clcb_progress)(size_t total_items, size_t now_completed, void *context);
+
+/**
+ * @brief Set a progress callback function to be called incrementally during a
+ * database load.
+ *
+ * Caution: changing options for an engine that is in-use is not thread-safe!
+ *
+ * @param engine    The initialized scanning engine
+ * @param callback  The callback function pointer
+ * @param context   Opaque application provided data
+ */
+extern void cl_engine_set_clcb_sigload_progress(struct cl_engine *engine, clcb_progress callback, void *context);
+
+/**
+ * @brief Set a progress callback function to be called incrementally during an
+ * engine compile.
+ *
+ * Disclaimer: the number of items for this is a rough estimate of the items that
+ * tend to take longest to compile and doesn't represent an accurate number of
+ * things compiled.
```

Comment:
So both `total_items` and `now_completed` will be a count of the slower items? (And is this true for _free_progress below as well?)

Reply:
Yes, the slower tasks, or just 1 task for a batch of frees. 

I actually haven't worked it into the bytecode signature compile feature though because that code is more complicated and would require adding parameters to a few functions. So if using LLVM then the last task is "compile all the bytecode sigs" and will take a hot second before the overall progress is complete.  

For what it's worth, using LLVM for the runtime in 0.104 isn't likely -- since upgrading to support a modern version of libLLVM missed the cut and will have to go into 0.105.  I would like to try to find a way to have the compile progress track each bytecode signature that we compile someday though.